_posts/2023-09-05-use-opensearch-benchmark-with-randomized-query-terms.md

```diff
@@ -0,0 +1,171 @@
+---
+layout: post
+title:  "Use OpenSearch Benchmark with Randomized Query Terms"
+authors:
+- handler
+date: 2023-09-05
+categories:
+  - technical-post
+meta_keywords: opensearch-benchmark
+meta_description: Learn how to add a randomized query set to an OpenSearch Benchmark workload
+---
+
+Performance testing is part science and part art. On the measurement side, you employ science to run test cases, collect and visualize metrics, and evaluate results. On the test creation side, you use intuition and understanding to make sure that you build coverage across the use cases and data that will comprise the majority of the uses of OpenSearch.
+
+[OpenSearch Benchmark](https://opensearch.org/docs/latest/benchmark/index/) is a community-driven, benchmarking framework that makes it easy to benchmark your workload against OpenSearch. You can use OpenSearch benchmark to create a custom workload, pulling data from your indices on OpenSearch, and then customize that workload to your requirements. 
+
+One limitation of OpenSearch Benchmark is that when you run a search operation, the query itself is hard-coded. If you want to randomize your query workload across a set of possible search terms, there's no out-of-the-box way to do that. In this post, you'll use data from [BoardgameGeek.com](https://boardgamegeek.com/) (BGG) to create a custom Opensearch Benchmark workload. You'll then create two, one-line-per-query files and use those files to create randomized terms for querying OpenSearch. 
+
+## Prerequisites
+
+To set up to follow this example, you’ll need the following:
+
+- **Python 3.8+**: Follow instructions to install [Python 3.8+](https://www.python.org/downloads/)
+- **Java Development Kit (JDK) 17**: Follow instructions to [install JDK 17](https://www.oracle.com/java/technologies/downloads/)
+- **OpenSearch Benchmark**: Follow instructions to [install OpenSearch Benchmark](https://opensearch.org/docs/latest/benchmark/installing-benchmark/). I installed in a virtual environment in a scratch directory that also holds the custom workload and config. 
+- **OpenSearch Cluster**: For this walkthrough, I used [Docker Desktop](https://www.docker.com/products/docker-desktop/) to run OpenSearch on my machine. Follow the [OpenSearch Quick Start](https://opensearch.org/docs/latest/quickstart/) to create an OpenSearch cluster.
+- **Data in OpenSearch**: [BoardgameGeek provides APIs](https://boardgamegeek.com/wiki/page/BGG_XML_API2) that deliver data to you in Extensible Markup Language (XML). For this example, I used and extended [sample code from the Open Distro for Elasticsearch project that downloads some games from BoardGameGeek](https://github.com/opendistro-for-elasticsearch/sample-code/tree/main/BGG) to load games into OpenSearch in a `games` index. The data contains names, descriptions, ratings, user comments, and more. If you have your own data set, you can load that into OpenSearch instead.
```

Comment:
Because BoardGameGeek has already been defined as BGG above, we can just use the abbreviation thereafter.